pkg/wekafs/controllerserver.go

```diff
@@ -191,81 +212,128 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
+	case apiclient.QuotaTypeHard:
+		enforceCapacity = true
+	case "":
+		enforceCapacity = false
+	default:
+		glog.Warningf("Could not recognize capacity enforcement in params: %s", qt)
+		return false, errors.New("unsupported capacityEnforcement in volume params")
+	}
+	return enforceCapacity, nil
+}
+
+func DeleteVolumeError(errorCode codes.Code, errorMessage string) (*csi.DeleteVolumeResponse, error) {
+	glog.Errorln("Error deleting volume, code:", errorCode, ", error:", errorMessage)
+	err := status.Error(errorCode, strings.ToLower(errorMessage))
+	return &csi.DeleteVolumeResponse{}, err
+}
+
 func (cs *controllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
 	volumeID := req.GetVolumeId()
 	if len(volumeID) == 0 {
 		return nil, status.Error(codes.InvalidArgument, "Volume ID missing in request")
 	}
 
-	volume, err := NewVolume(volumeID)
+	client, err := cs.api.GetClientFromSecrets(req.Secrets)
+	if err != nil {
+		return DeleteVolumeError(codes.Internal, "Failed to initialize Weka API client for the request")
+	}
+
+	volume, err := NewVolume(volumeID, client)
 	if err != nil {
-		return &csi.DeleteVolumeResponse{}, nil // Should return invalid on incorrect ID
+		// Should return ok on incorrect ID (by CSI spec)
+		return &csi.DeleteVolumeResponse{}, nil
 	}
 
 	if err := cs.validateControllerServiceRequest(csi.ControllerServiceCapability_RPC_CREATE_DELETE_VOLUME); err != nil {
 		glog.V(3).Infof("invalid delete volume req: %v", req)
-		return nil, err
+		return DeleteVolumeError(codes.Internal, err.Error())
 	}
 
-	// Perform mount in order to be able to access Xattrs and get a full volume root path
-	glog.V(4).Infof("Deleting volume %s, located in filesystem %s", volume.id, volume.fs)
-
 	err = volume.moveToTrash(cs.mounter, cs.gc)
 	if os.IsNotExist(err) {
-		glog.V(4).Infof("Volume not found %s, but returning success for idempotence", volume.id)
+		glog.V(4).Infof("Volume not found %s, but returning success for idempotence", volume.GetId())
 		return &csi.DeleteVolumeResponse{}, nil
 	}
-	return &csi.DeleteVolumeResponse{}, err
+	// cleanup
+	if err != nil {
+		return DeleteVolumeError(codes.Internal, err.Error())
+	}
+	return &csi.DeleteVolumeResponse{}, nil
+}
+
+func ExpandVolumeError(errorCode codes.Code, errorMessage string) (*csi.ControllerExpandVolumeResponse, error) {
+	glog.Errorln("Error expanding volume, code:", errorCode, ", error:", errorMessage)
+	err := status.Error(errorCode, strings.ToLower(errorMessage))
+	return &csi.ControllerExpandVolumeResponse{}, err
 }
 
 func (cs *controllerServer) ControllerExpandVolume(ctx context.Context, req *csi.ControllerExpandVolumeRequest) (*csi.ControllerExpandVolumeResponse, error) {
 
 	if len(req.GetVolumeId()) == 0 {
 		return nil, status.Errorf(codes.InvalidArgument, "Volume ID not specified")
 	}
-	volume, err := NewVolume(req.GetVolumeId())
+	client, err := cs.api.GetClientFromSecrets(req.Secrets)
```

Comment:
need to check if this is not causing issues with backward compatibility, e.g. 
delete still works for volumes without API

Reply:
Apparently not, since in this case we `get client, err := nil, nil`